File: packages/python/src/geographicol/parse.py
"""The parsing procedure from docs/nomenclature.md section 12, step by step."""

from __future__ import annotations

from .codes import (
    AMBIGUOUS_COMPLEMENT_ALIASES,
    CALLE_LIKE,
    CARRERA_LIKE,
    COMPLEMENTS,
    NUMBER_MARKERS,
    ONCE_ONLY,
    PENALTIES,
    QUADRANT_ALIASES,
    SINGLE_LETTER_QUADRANTS,
    STREET_ALIASES,
    UNCOMMON_QUADRANTS,
    ComplementCode,
)
from .render import render
from .tokenize import Token, tokenize
from .types import Complement, ParsedAddress, Quadrant, StreetType, Style

# Street aliases as token lists, longest first so "avenida carrera" wins over "avenida".
_STREET_MATCHERS: list[tuple[StreetType, list[str]]] = sorted(
    (
        (street_type, alias.split(" "))
        for street_type, aliases in STREET_ALIASES.items()
        for alias in aliases
    ),
    key=lambda m: -len(m[1]),
)

_COMPLEMENT_MATCHERS: list[tuple[ComplementCode, list[str]]] = sorted(
    ((code, alias.split(" ")) for code in COMPLEMENTS for alias in code.aliases),
    key=lambda m: -len(m[1]),
)


class _AddressParser:
    def __init__(self, raw: str) -> None:
        self.raw = raw
        self.tokens: list[Token] = []
        self.i = 0
        self.a = ParsedAddress(raw=raw)
        self.raised: list[str] = []
        """Every warning occurrence, in order. Penalties are counted from this list."""

    def run(self) -> None:
        segments = self.raw.split(",")
        self._parse_address_segment(segments[0])
        for segment in segments[1:]:
            self._parse_extra_segment(segment)

        if self.a.locality is not None or self.a.department is not None:
            self._warn("LOCALITY_UNVERIFIED")

    def _warn(self, code: str) -> None:
        self.raised.append(code)

    def _peek(self, offset: int = 0) -> Token | None:
        index = self.i + offset
        return self.tokens[index] if index < len(self.tokens) else None

    # Steps 5 to 9, on the text before the first comma.
    def _parse_address_segment(self, text: str) -> None:
        self.tokens = tokenize(text)
        self.i = 0
        a = self.a

        a.street_type = self._parse_street_type()
        if a.street_type is None:
            self._warn("UNRECOGNIZED_STREET_TYPE")

        if a.street_type is not None and self._is_named_street():
            a.street_name = self._read_street_name()
            self._warn("NAMED_STREET")
        else:
            a.street_number = self._parse_number(3)
            if a.street_number is not None:
                a.street_letter = self._parse_suffix()
                a.street_quadrant = self._parse_quadrant()
        if a.street_number is None and a.street_name is None:
            self._warn("MISSING_STREET_NUMBER")

        if a.street_type == "KM":
            self._warn("RURAL_ADDRESS")
            # A KM address is a road location: the rest is the road description.
            rest = self._peek()
            if rest:
                a.locality = text[rest.start :].strip()
            self.i = len(self.tokens)
        else:
            self._parse_placa()

        if a.cross_number is None:
            self._warn("MISSING_CROSS_NUMBER")
        elif a.plate_number is None:
            self._warn("MISSING_PLATE_NUMBER")

        self._parse_complements_and_leftovers(text, locality_allowed=True)

    # Step 3: a later comma segment is complements, or else locality, then department.
    def _parse_extra_segment(self, text: str) -> None:
        self.tokens = tokenize(text)
        self.i = 0
        if not self.tokens:
            return
        if self._match_complement():
            self._parse_complements_and_leftovers(text, locality_allowed=False)
            return
        value = text.strip()
        if self.a.locality is None:
            self.a.locality = value
        elif self.a.department is None:
            self.a.department = value
        else:
            for _ in self.tokens:
                self._warn("UNKNOWN_TOKEN")

    # Step 5.
    def _parse_street_type(self) -> StreetType | None:
        for street_type, words in _STREET_MATCHERS:
            if self._matches_words(words):
                self.i += len(words)
                if len(words) == 1 and len(words[0]) == 1:
                    self._warn("AMBIGUOUS_STREET_TYPE")
                return street_type
        return None

    def _matches_words(self, words: list[str]) -> bool:
        for k, word in enumerate(words):
            token = self._peek(k)
            if token is None or token.is_num or token.norm != word:
                return False
        return True

    @staticmethod
    def _is_number_marker(token: Token | None) -> bool:
        return token is not None and not token.is_num and token.norm in NUMBER_MARKERS

    def _is_named_street(self) -> bool:
        """A word right after the street type, other than a number marker, starts a street name."""
        token = self._peek()
        return token is not None and not token.is_num and not self._is_number_marker(token)

    def _read_street_name(self) -> str:
        """Words up to the first number or number marker: "BOYACÁ", "DE LA FACTORÍA"."""
        words: list[str] = []
        while (token := self._peek()) and not token.is_num and not self._is_number_marker(token):
            words.append(token.orig)
            self.i += 1
        return " ".join(words).upper()

    def _parse_number(self, max_digits: int) -> int | None:
        token = self._peek()
        if token is None or not token.is_num or len(token.norm) > max_digits:
            return None
        self.i += 1
        return int(token.norm)

    # Letters and BIS after a number (section 4).
    def _parse_suffix(self) -> str | None:
        parts: list[str] = []
        while (token := self._peek()) and not token.is_num:
            if token.norm == "bis":
                parts.append("BIS")
            elif len(token.norm) == 1 and "a" <= token.norm <= "z":
                # N is never a letter: attached it is NORTE, alone it means "número".
                # Standing alone, S and E are quadrants.
                if token.norm == "n":
                    break
                if not token.attached and token.norm in SINGLE_LETTER_QUADRANTS:
                    break
                parts.append(token.norm.upper())
            else:
                break
            self.i += 1
        return " ".join(parts) if parts else None

    # Section 5.
    def _parse_quadrant(self) -> Quadrant | None:
        token = self._peek()
        if token is None or token.is_num:
            return None
        quadrant = QUADRANT_ALIASES.get(token.norm)
        if quadrant is None and not token.attached:
            quadrant = SINGLE_LETTER_QUADRANTS.get(token.norm)
            if quadrant:
                self._warn("AMBIGUOUS_QUADRANT")
        if quadrant is None and token.attached and token.norm == "n":
            # Cali's "6N": an attached N is NORTE.
            quadrant = "NORTE"
            self._warn("AMBIGUOUS_QUADRANT")
        if quadrant is None:
            return None
        if quadrant in UNCOMMON_QUADRANTS:
            self._warn("UNCOMMON_QUADRANT")
        self.i += 1
        return quadrant

    # Steps 6-7: number markers, cross street, plate (section 6).
    def _parse_placa(self) -> None:
        a = self.a
        while self._is_number_marker(self._peek()):
            self.i += 1

        block = self._peek()
        if block is None or not block.is_num:
            return
        if len(block.norm) >= 5:
            return  # left for the leftovers step: UNKNOWN_TOKEN
        self.i += 1
        a.cross_letter = self._parse_suffix()
        a.cross_quadrant = self._parse_quadrant()

        plate = self._peek()
        plate_follows = plate is not None and plate.is_num and len(plate.norm) <= 3
        if (
            not plate_follows
            and a.cross_letter is None
            and a.cross_quadrant is None
            and len(block.norm) >= 3
        ):
            # Glued cross and plate: "1230" -> 12 and 30.
            a.cross_number = int(block.norm[:-2])
            a.plate_number = int(block.norm[-2:])
            self._warn("AMBIGUOUS_PLATE")
        else:
            a.cross_number = int(block.norm)
            if plate_follows and plate is not None:
                self.i += 1
                a.plate_number = int(plate.norm)
        if a.cross_quadrant is None:
            self._parse_trailing_quadrant()

    def _parse_trailing_quadrant(self) -> None:
        """A quadrant after the plate follows the kind of street it describes (section 5)."""
        a = self.a
        quadrant = self._parse_quadrant()
        if quadrant is None:
            return
        street = a.street_type
        belongs_to_street = (
            a.street_quadrant is None
            and street is not None
            and (
                (quadrant == "SUR" and street in CALLE_LIKE)
                or (quadrant == "ESTE" and street in CARRERA_LIKE)
            )
        )
        if belongs_to_street:
            a.street_quadrant = quadrant
        else:
            a.cross_quadrant = quadrant

    def _match_complement(self) -> tuple[ComplementCode, int, str] | None:
        for code, words in _COMPLEMENT_MATCHERS:
            if not self._matches_words(words):
                continue
            alias = " ".join(words)
            following = self._peek(len(words))
            # "tr" is Torre only once the placa is parsed; "p" and "l" only before a number.
            if alias == "tr" and self.a.cross_number is None:
                continue
            if alias in ("p", "l") and (following is None or not following.is_num):
                continue
            return code, len(words), alias
        return None

    # Steps 8-9.
    def _parse_complements_and_leftovers(self, text: str, locality_allowed: bool) -> None:
        while self._peek():
            match = self._match_complement()
            if match:
                code, length, alias = match
                self.i += length
                if alias in AMBIGUOUS_COMPLEMENT_ALIASES:
                    self._warn("AMBIGUOUS_COMPLEMENT")
                value = self._read_name_value() if code.name_valued else self._read_token_value()
                if value is None:
                    self._warn("UNKNOWN_TOKEN")
                    continue
                self.a.complements.append(
                    Complement(type=code.type, code=code.catastral, value=value)
                )
                continue

            rest = self.tokens[self.i :]
            all_words = all(not token.is_num for token in rest)
            if (
                locality_allowed
                and all_words
                and self.a.cross_number is not None
                and self.a.locality is None
            ):
                self.a.locality = text[rest[0].start :].strip()
                self.i = len(self.tokens)
                return
            self._warn("UNKNOWN_TOKEN")
            self.i += 1

    def _read_token_value(self) -> str | None:
        """One token, plus any tokens attached to it: "501B"."""
        first = self._peek()
        if first is None or self._match_complement():
            return None
        value = first.orig
        self.i += 1
        while (token := self._peek()) and token.attached:
            value += token.orig
            self.i += 1
        return value.upper()

    def _read_name_value(self) -> str | None:
        """Words up to the next complement keyword or the end: "SAN FERNANDO"."""
        words: list[str] = []
        while (token := self._peek()) and not self._match_complement():
            words.append(token.orig)
            self.i += 1
        return " ".join(words).upper() if words else None


def _score(raised: list[str], strict: bool) -> float:
    penalty = 0.0
    counted: set[str] = set()
    for code in raised:
        if code in ONCE_ONLY and code in counted:
            continue
        counted.add(code)
        penalty += PENALTIES.get(code, 0)
    if strict:
        penalty *= 2
    return min(1.0, max(0.0, 1 - penalty))


def parse(input: str, style: Style = "catastral", strict: bool = False) -> ParsedAddress:
    """Parses a Colombian address.

    Never raises: unparseable input gets confidence 0 and warnings instead.
    """
    if input.strip() == "":
        return ParsedAddress(warnings=["EMPTY_INPUT"], raw=input)

    parser = _AddressParser(input)
    parser.run()
    address = parser.a
    address.canonical = render(address, style)
    address.normalized = render(address, "readable")
    address.confidence = _score(parser.raised, strict)
    address.warnings = list(dict.fromkeys(parser.raised))
    return address

File: packages/node/src/parse.ts
// The parsing procedure from docs/nomenclature.md section 12, step by step.

import {
  AMBIGUOUS_COMPLEMENT_ALIASES,
  CALLE_LIKE,
  CARRERA_LIKE,
  COMPLEMENTS,
  type ComplementCode,
  NUMBER_MARKERS,
  ONCE_ONLY,
  PENALTIES,
  QUADRANT_ALIASES,
  SINGLE_LETTER_QUADRANTS,
  STREET_ALIASES,
  UNCOMMON_QUADRANTS,
} from "./codes.js";
import { render } from "./render.js";
import { type Token, tokenize } from "./tokenize.js";
import type { Complement, ParseOptions, ParsedAddress, Quadrant, StreetType } from "./types.js";

/** Street aliases as token lists, longest first so "avenida carrera" wins over "avenida". */
const STREET_MATCHERS: Array<{ type: StreetType; words: string[] }> = Object.entries(STREET_ALIASES)
  .flatMap(([type, aliases]) =>
    aliases.map((alias) => ({ type: type as StreetType, words: alias.split(" ") })),
  )
  .sort((a, b) => b.words.length - a.words.length);

const COMPLEMENT_MATCHERS: Array<{ code: ComplementCode; words: string[] }> = COMPLEMENTS.flatMap(
  (code) => code.aliases.map((alias) => ({ code, words: alias.split(" ") })),
).sort((a, b) => b.words.length - a.words.length);

type Fields = Omit<ParsedAddress, "canonical" | "normalized" | "confidence" | "warnings" | "raw">;

function emptyFields(): Fields {
  return {
    streetType: null,
    streetNumber: null,
    streetName: null,
    streetLetter: null,
    streetQuadrant: null,
    crossNumber: null,
    crossLetter: null,
    crossQuadrant: null,
    plateNumber: null,
    complements: [],
    locality: null,
    department: null,
  };
}

class AddressParser {
  private tokens: Token[] = [];
  private i = 0;
  readonly fields = emptyFields();
  /** Every warning occurrence, in order. Penalties are counted from this list. */
  readonly raised: string[] = [];

  constructor(private readonly raw: string) {}

  run(): void {
    const segments = this.raw.split(",");
    this.parseAddressSegment(segments[0] ?? "");
    for (const segment of segments.slice(1)) this.parseExtraSegment(segment);

    if (this.fields.locality !== null || this.fields.department !== null) {
      this.warn("LOCALITY_UNVERIFIED");
    }
  }

  private warn(code: string): void {
    this.raised.push(code);
  }

  private peek(offset = 0): Token | undefined {
    return this.tokens[this.i + offset];
  }

  // Steps 5 to 9, on the text before the first comma.
  private parseAddressSegment(text: string): void {
    this.tokens = tokenize(text);
    this.i = 0;
    const f = this.fields;

    f.streetType = this.parseStreetType();
    if (f.streetType === null) this.warn("UNRECOGNIZED_STREET_TYPE");

    if (f.streetType !== null && this.isNamedStreet()) {
      f.streetName = this.readStreetName();
      this.warn("NAMED_STREET");
    } else {
      f.streetNumber = this.parseNumber(3);
      if (f.streetNumber !== null) {
        f.streetLetter = this.parseSuffix();
        f.streetQuadrant = this.parseQuadrant();
      }
    }
    if (f.streetNumber === null && f.streetName === null) this.warn("MISSING_STREET_NUMBER");

    if (f.streetType === "KM") {
      this.warn("RURAL_ADDRESS");
      // A KM address is a road location: the rest is the road description.
      const rest = this.peek();
      if (rest) f.locality = text.slice(rest.start).trim();
      this.i = this.tokens.length;
    } else {
      this.parsePlaca();
    }

    if (f.crossNumber === null) this.warn("MISSING_CROSS_NUMBER");
    else if (f.plateNumber === null) this.warn("MISSING_PLATE_NUMBER");

    this.parseComplementsAndLeftovers(text, true);
  }

  // Step 3: a later comma segment is complements, or else locality, then department.
  private parseExtraSegment(text: string): void {
    this.tokens = tokenize(text);
    this.i = 0;
    if (this.tokens.length === 0) return;
    if (this.matchComplement()) {
      this.parseComplementsAndLeftovers(text, false);
      return;
    }
    const value = text.trim();
    if (this.fields.locality === null) this.fields.locality = value;
    else if (this.fields.department === null) this.fields.department = value;
    else for (const _ of this.tokens) this.warn("UNKNOWN_TOKEN");
  }

  // Step 5.
  private parseStreetType(): StreetType | null {
    for (const matcher of STREET_MATCHERS) {
      if (this.matchesWords(matcher.words)) {
        this.i += matcher.words.length;
        if (matcher.words.length === 1 && (matcher.words[0] ?? "").length === 1) {
          this.warn("AMBIGUOUS_STREET_TYPE");
        }
        return matcher.type;
      }
    }
    return null;
  }

  private matchesWords(words: string[]): boolean {
    return words.every((word, k) => {
      const token = this.peek(k);
      return token !== undefined && !token.isNum && token.norm === word;
    });
  }

  private isNumberMarker(token: Token | undefined): boolean {
    return token !== undefined && !token.isNum && NUMBER_MARKERS.includes(token.norm);
  }

  /** A word right after the street type, other than a number marker, starts a street name. */
  private isNamedStreet(): boolean {
    const next = this.peek();
    return next !== undefined && !next.isNum && !this.isNumberMarker(next);
  }

  /** Words up to the first number or number marker: "BOYACÁ", "DE LA FACTORÍA". */
  private readStreetName(): string {
    const words: string[] = [];
    while (this.peek() && !this.peek()?.isNum && !this.isNumberMarker(this.peek())) {
      words.push(this.peek()?.orig ?? "");
      this.i++;
    }
    return words.join(" ").toUpperCase();
  }

  private parseNumber(maxDigits: number): number | null {
    const token = this.peek();
    if (!token?.isNum || token.norm.length > maxDigits) return null;
    this.i++;
    return Number(token.norm);
  }

  // Letters and BIS after a number (section 4).
  private parseSuffix(): string | null {
    const parts: string[] = [];
    for (;;) {
      const token = this.peek();
      if (!token || token.isNum) break;
      if (token.norm === "bis") {
        parts.push("BIS");
      } else if (token.norm.length === 1 && /[a-z]/.test(token.norm)) {
        // N is never a letter: attached it is NORTE, alone it means "número".
        // Standing alone, S and E are quadrants.
        if (token.norm === "n") break;
        if (!token.attached && token.norm in SINGLE_LETTER_QUADRANTS) break;
        parts.push(token.norm.toUpperCase());
      } else {
        break;
      }
      this.i++;
    }
    return parts.length > 0 ? parts.join(" ") : null;
  }

  // Section 5.
  private parseQuadrant(): Quadrant | null {
    const token = this.peek();
    if (!token || token.isNum) return null;
    let quadrant = QUADRANT_ALIASES[token.norm];
    if (!quadrant && !token.attached) {
      quadrant = SINGLE_LETTER_QUADRANTS[token.norm];
      if (quadrant) this.warn("AMBIGUOUS_QUADRANT");
    }
    if (!quadrant && token.attached && token.norm === "n") {
      // Cali's "6N": an attached N is NORTE.
      quadrant = "NORTE";
      this.warn("AMBIGUOUS_QUADRANT");
    }
    if (!quadrant) return null;
    if (UNCOMMON_QUADRANTS.includes(quadrant)) this.warn("UNCOMMON_QUADRANT");
    this.i++;
    return quadrant;
  }

  // Steps 6-7: number markers, cross street, plate (section 6).
  private parsePlaca(): void {
    const f = this.fields;
    while (this.isNumberMarker(this.peek())) this.i++;

    const block = this.peek();
    if (!block?.isNum) return;
    if (block.norm.length >= 5) return; // left for the leftovers step: UNKNOWN_TOKEN
    this.i++;
    f.crossLetter = this.parseSuffix();
    f.crossQuadrant = this.parseQuadrant();

    const plate = this.peek();
    const plateFollows = plate?.isNum === true && plate.norm.length <= 3;
    if (
      !plateFollows &&
      f.crossLetter === null &&
      f.crossQuadrant === null &&
      block.norm.length >= 3
    ) {
      // Glued cross and plate: "1230" -> 12 and 30.
      f.crossNumber = Number(block.norm.slice(0, -2));
      f.plateNumber = Number(block.norm.slice(-2));
      this.warn("AMBIGUOUS_PLATE");
    } else {
      f.crossNumber = Number(block.norm);
      if (plateFollows) {
        this.i++;
        f.plateNumber = Number(plate.norm);
      }
    }
    if (f.crossQuadrant === null) this.parseTrailingQuadrant();
  }

  /** A quadrant after the plate follows the kind of street it describes (section 5). */
  private parseTrailingQuadrant(): void {
    const f = this.fields;
    const quadrant = this.parseQuadrant();
    if (quadrant === null) return;
    const street = f.streetType;
    const belongsToStreet =
      f.streetQuadrant === null &&
      street !== null &&
      ((quadrant === "SUR" && CALLE_LIKE.includes(street)) ||
        (quadrant === "ESTE" && CARRERA_LIKE.includes(street)));
    if (belongsToStreet) f.streetQuadrant = quadrant;
    else f.crossQuadrant = quadrant;
  }

  private matchComplement(): { code: ComplementCode; length: number; alias: string } | null {
    for (const matcher of COMPLEMENT_MATCHERS) {
      if (!this.matchesWords(matcher.words)) continue;
      const alias = matcher.words.join(" ");
      const next = this.peek(matcher.words.length);
      // "tr" is Torre only once the placa is parsed; "p" and "l" only before a number.
      if (alias === "tr" && this.fields.crossNumber === null) continue;
      if ((alias === "p" || alias === "l") && !next?.isNum) continue;
      return { code: matcher.code, length: matcher.words.length, alias };
    }
    return null;
  }

  // Steps 8-9.
  private parseComplementsAndLeftovers(text: string, localityAllowed: boolean): void {
    while (this.peek()) {
      const match = this.matchComplement();
      if (match) {
        this.i += match.length;
        if (AMBIGUOUS_COMPLEMENT_ALIASES.includes(match.alias)) this.warn("AMBIGUOUS_COMPLEMENT");
        const value = match.code.nameValued ? this.readNameValue() : this.readTokenValue();
        if (value === null) {
          this.warn("UNKNOWN_TOKEN");
          continue;
        }
        const complement: Complement = { type: match.code.type, code: match.code.catastral, value };
        this.fields.complements.push(complement);
        continue;
      }

      const rest = this.tokens.slice(this.i);
      const allWords = rest.every((token) => !token.isNum);
      if (
        localityAllowed &&
        allWords &&
        this.fields.crossNumber !== null &&
        this.fields.locality === null
      ) {
        this.fields.locality = text.slice(rest[0]?.start ?? 0).trim();
        this.i = this.tokens.length;
        return;
      }
      this.warn("UNKNOWN_TOKEN");
      this.i++;
    }
  }

  /** One token, plus any tokens attached to it: "501B". */
  private readTokenValue(): string | null {
    const first = this.peek();
    if (!first || this.matchComplement()) return null;
    let value = first.orig;
    this.i++;
    while (this.peek()?.attached) {
      value += this.peek()?.orig ?? "";
      this.i++;
    }
    return value.toUpperCase();
  }

  /** Words up to the next complement keyword or the end: "SAN FERNANDO". */
  private readNameValue(): string | null {
    const words: string[] = [];
    while (this.peek() && !this.matchComplement()) {
      words.push(this.peek()?.orig ?? "");
      this.i++;
    }
    return words.length > 0 ? words.join(" ").toUpperCase() : null;
  }
}

function score(raised: string[], strict: boolean): number {
  let penalty = 0;
  const counted = new Set<string>();
  for (const code of raised) {
    if (ONCE_ONLY.includes(code) && counted.has(code)) continue;
    counted.add(code);
    penalty += PENALTIES[code] ?? 0;
  }
  if (strict) penalty *= 2;
  return Math.min(1, Math.max(0, 1 - penalty));
}

export function parse(input: string, options: ParseOptions = {}): ParsedAddress {
  const raw = input;
  if (raw.trim() === "") {
    const fields = emptyFields();
    return {
      ...fields,
      canonical: "",
      normalized: "",
      confidence: 0,
      warnings: ["EMPTY_INPUT"],
      raw,
    };
  }

  const parser = new AddressParser(raw);
  parser.run();
  const fields = parser.fields;
  return {
    ...fields,
    canonical: render(fields, options.style ?? "catastral"),
    normalized: render(fields, "readable"),
    confidence: score(parser.raised, options.strict ?? false),
    warnings: [...new Set(parser.raised)],
    raw,
  };
}
